glean_parser/templates/python_server.jinja2

```diff
@@ -0,0 +1,186 @@
+
+
+{% for ping, metrics_by_type in pings.items() %}
+class {{ ping|camelize }}ServerEventLogger:
+    def __init__(
+        self, application_id: str, app_display_version: str, channel: str
+    ) -> None:
+        """
+        Create {{ ping|camelize }}ServerEventLogger instance.
+
+        :param str application_id: The application ID.
+        :param str app_display_version: The application display version.
+        :param str channel: The channel.
+        """
+        self._application_id = application_id
+        self._app_display_version = app_display_version
+        self._channel = channel
+
+    def _record(
+		self,
+		user_agent: str,
+		ip_address: str,
+		{% for metric_type, metrics in metrics_by_type.items() %}
+		{% if metric_type != 'event' %}
+		{% for metric in metrics %}
+		{{ metric.category }}_{{ metric.name }}: {{ metric.type|py_metric_type }},
+		{% endfor %}
+		{% endif %}
+		{% endfor %}
+		event: dict[str, Any]
+	) -> None:
+        now = datetime.now(timezone.utc)
+        timestamp = now.isoformat()
+        event["timestamp"] = int(1000.0 * now.timestamp())  # Milliseconds since epoch
+        event_payload = {
+            "metrics": {
+				{% for metric_type, metrics in metrics_by_type.items() %}
+				{% if metric_type != 'event' %}
+				"{{ metric_type }}": {
+					{% for metric in metrics %}
+					"{{ metric.category }}.{{ metric.name }}": {{ metric.category }}_{{ metric.name }},
+					{% endfor %}
+				},
+				{% endif %}
+				{% endfor %}
+			},
+            "events": [event],
+            "ping_info": {
+                # seq is required in the Glean schema, however is not useful in server context
+                "seq": 0,
+                "start_time": timestamp,
+                "end_time": timestamp,
+            },
+            # `Unknown` fields below are required in the Glean schema, however they are
+            # not useful in server context
+            "client_info": {
+                "telemetry_sdk_build": "glean_parser v{{ parser_version }}",
+                "first_run_date": "Unknown",
+                "os": "Unknown",
+                "os_version": "Unknown",
+                "architecture": "Unknown",
+                "app_build": "Unknown",
+                "app_display_version": self._app_display_version,
+                "app_channel": self._channel,
+            },
+        }
+        event_payload_serialized = json.dumps(event_payload)
+
+        # This is the message structure that Decoder expects:
+        # https://github.com/mozilla/gcp-ingestion/pull/2400
+        ping = {
+            "document_namespace": self._application_id,
+            "document_type": "{{ ping }}",
+            "document_version": "1",
+            "document_id": str(uuid4()),
+            "user_agent": user_agent,
+            "ip_address": ip_address,
+            "payload": event_payload_serialized,
+        }
+
+        ping_envelope = {
+            "Timestamp": timestamp,
```

Comment:
Our existing logs uses integer nanoseconds since the epoch (`time.time_ns`), not the ISO format. Does this get consumed by the Glean ingester, or is it just part of the envelope?

Reply:
It's not consumed, see also https://github.com/mozilla/glean_parser/pull/653#discussion_r1471524494